File: prototype_abstract/src/smart_belt/drag.rs
use super::{Error, RaySense, SmartBeltWorldView};
use super::drag_state::{DragStepResult, LastBuiltEntity, step};
use crate::belts::BeltTier;
use crate::world::WorldImpl;
use crate::{BeltConnectable, BeltConnections, TilePosition};
use crate::{Direction, Ray};
use log::debug;

/// Handles dragging in a straight line (no rotations).
pub struct LineDrag<'a> {
    pub(super) world: &'a mut WorldImpl,
    pub(super) ray: Ray,
    pub(super) tier: BeltTier,
    last_position: i32,
    // Some tiles we just placed may change other belt's curvature; however we
    // want the logic to be independent of what we've placed. As such, we track
    // the history of tiles we've replaced. It suffices only to keep track of
    // one tile (the last placed output belt).
    // See world_view.rs for more info
    tile_history: Option<BeltConnections>,
    // After rotation, we also may want the last placed belt of the previous rotation in tile history.
    last_end_tile_history: Option<(TilePosition, BeltConnections)>,

    pub(super) forward_placement: i32,
    pub(super) backward_placement: i32,
    pub(super) furthest_placement_direction: RaySense,

    pub(super) forward_pos: i32,
    pub(super) backward_pos: i32,
    pub(super) rotation_pivot_direction: RaySense,

    pub(super) last_built_entity: Option<LastBuiltEntity>,
    pub(super) over_impassable: Option<RaySense>,
}

impl<'a> LineDrag<'a> {
    /// Starts a drag.
    /// The very first click may fast-replace something, forcing something to be overwritten.
    fn new_drag(
        world: &'a mut WorldImpl,
        error_handler: &mut dyn FnMut(TilePosition, Error),
        tier: BeltTier,
        start_pos: TilePosition,
        belt_direction: Direction,
        first_belt_direction: Direction,
        allow_fast_replace: bool,
    ) -> LineDrag<'a> {
        let can_place =
            world.can_place_or_fast_replace_belt(start_pos, belt_direction, allow_fast_replace);
        let tile_history = can_place.then(|| world.belt_connections_at(start_pos));

        let ray = Ray::new(start_pos, belt_direction);
        let start_coord = ray.ray_position(start_pos);

        let last_built_entity = if can_place {
            let entity = world.place_belt(start_pos, first_belt_direction, tier);
            let connectable = BeltConnectable::try_from(entity).unwrap();
            Some(LastBuiltEntity::from_build(connectable, start_coord))
        } else {
            error_handler(start_pos, Error::EntityInTheWay);
            None
        };

        LineDrag {
            world,
            ray,
            tier,
            last_position: start_coord,
            tile_history,
            last_end_tile_history: None,
            forward_placement: start_coord,
            backward_placement: start_coord,
            furthest_placement_direction: RaySense::Forward,
            forward_pos: start_coord,
            backward_pos: start_coord,
            rotation_pivot_direction: RaySense::Forward,
            last_built_entity,
            over_impassable: None,
        }
    }

    /// Starts a drag.
    /// The very first click may fast-replace something, forcing something to be overwritten.
    pub fn start_drag(
        world: &'a mut WorldImpl,
        error_handler: &mut dyn FnMut(TilePosition, Error),
        tier: BeltTier,
        start_pos: TilePosition,
        belt_direction: Direction,
    ) -> LineDrag<'a> {
        Self::new_drag(
            world,
            error_handler,
            tier,
            start_pos,
            belt_direction,
            belt_direction,
            true,
        )
    }

    pub fn rotate(
        self,
        error_handler: &mut dyn FnMut(TilePosition, Error),
        cursor_pos: TilePosition,
    ) -> (Self, bool) {
        let turn_direction = match self.ray.relative_direction(cursor_pos) {
            Some(dir) => dir,
            None => {
                return (self, false);
            }
        };

        let (pivot, backward) = self.get_rotation_pivot();
        let old_direction = self.ray.direction;

        let (new_belt_direction, first_belt_direction) = if backward {
            (turn_direction.opposite(), old_direction)
        } else {
            (turn_direction, turn_direction)
        };

        let last_tile_history = if !backward
            && let Some(connections) = self.tile_history
            && self.furthest_placement_pos() == self.last_position
        {
            let world_pos = self.ray.get_position(self.last_position);
            Some((world_pos, connections))
        } else {
            None
        };

        let mut new_line_drag = LineDrag::new_drag(
            self.world,
            error_handler,
            self.tier,
            pivot,
            new_belt_direction,
            first_belt_direction,
            false,
        );
        new_line_drag.last_end_tile_history = last_tile_history;
        new_line_drag.interpolate_to(error_handler, cursor_pos);

        (new_line_drag, true)
    }

    /// Main entry point for the drag operation.
    pub fn interpolate_to(
        &mut self,
        error_handler: &mut dyn FnMut(TilePosition, Error),
        new_position: TilePosition,
    ) {
        let target_pos = self.ray.ray_position(new_position);
        while self.ray.is_before(self.last_position, target_pos) {
            let view = self.create_world_view(RaySense::Forward);
            let result = step(self.last_built_entity.as_ref(), self.over_impassable, &view);
            self.apply_step(error_handler, result, RaySense::Forward);
        }
        while self.ray.is_before(target_pos, self.last_position) {
            let view = self.create_world_view(RaySense::Backward);
            let result = step(self.last_built_entity.as_ref(), self.over_impassable, &view);
            self.apply_step(error_handler, result, RaySense::Backward);
        }
        self.update_furthest_position(target_pos);
    }

    fn apply_step(
        &mut self,
        error_handler: &mut dyn FnMut(TilePosition, Error),
        result: DragStepResult,
        ray_sense: RaySense,
    ) {
        let DragStepResult(action, error) = result;
        debug!("action: {:?}", action);
        let next_position = self.last_position
            + self.ray.direction.axis_sign() * ray_sense.direction_multiplier();
        if action.is_placement() {
            self.update_furthest_placement(next_position, ray_sense)
        }
        self.apply_action(error_handler, action, next_position, ray_sense);

        if let Some(error) = error {
            let world_pos = self.ray.get_position(next_position);
            debug!("error: {:?}", error);
            error_handler(world_pos, error);
        }

        self.last_position = next_position;
    }

    fn store_tile_history(&mut self, position: i32) {
        let world_pos = self.ray.get_position(position);
        let tile_history = self.world.belt_connections_at(world_pos);
        self.tile_history = Some(tile_history);
        debug!("New tile history: {:?}", self.tile_history);
    }

    fn update_furthest_placement(&mut self, position: i32, ray_sense: RaySense) {
        match ray_sense {
            RaySense::Forward => {
                if self.ray.is_before(self.forward_placement, position) {
                    self.forward_placement = position;
                    self.store_tile_history(position);
                    self.furthest_placement_direction = RaySense::Forward;
                }
            }
            RaySense::Backward => {
                if self.ray.is_before(position, self.backward_placement) {
                    self.backward_placement = position;
                    self.store_tile_history(position);
                    self.furthest_placement_direction = RaySense::Backward;
                }
            }
        };
    }

    fn update_furthest_position(&mut self, target_pos: i32) {
        if self.ray.is_before(self.forward_pos, target_pos) {
            self.forward_pos = target_pos;
            self.rotation_pivot_direction = RaySense::Forward;
        }
        if self.ray.is_before(target_pos, self.backward_pos) {
            self.backward_pos = target_pos;
            self.rotation_pivot_direction = RaySense::Backward;
        }
    }

    pub(super) fn get_rotation_pivot(&self) -> (TilePosition, bool) {
        let furthest_pos = match self.rotation_pivot_direction {
            RaySense::Forward => self.forward_pos,
            RaySense::Backward => self.backward_pos,
        };
        (
            self.ray.get_position(furthest_pos),
            self.rotation_pivot_direction == RaySense::Backward,
        )
    }

    pub fn furthest_placement_pos(&self) -> i32 {
        match self.furthest_placement_direction {
            RaySense::Forward => self.forward_placement,
            RaySense::Backward => self.backward_placement,
        }
    }

    #[inline]
    pub(super) fn create_world_view(&self, ray_sense: RaySense) -> SmartBeltWorldView<'_> {
        let tile_history = self
            .tile_history
            .map(|h| (self.ray.get_position(self.furthest_placement_pos()), h))
            .into_iter()
            .chain(self.last_end_tile_history)
            .collect::<Vec<_>>();
        SmartBeltWorldView {
            world: self.world,
            ray: self.ray,
            tier: self.tier,
            last_position: self.last_position,
            tile_history,
            ray_sense,
        }
    }

    pub(super) fn report_error(
        error_handler: &mut dyn FnMut(TilePosition, Error),
        error: Error,
        world_pos: TilePosition,
    ) {
        debug!("error: {:?}", error);
        error_handler(world_pos, error);
    }
}

File: prototype_mod/common/smart_belt/drag.ts
import {
  BeltConnectable,
  Splitter,
  UndergroundBelt,
  type BeltTier,
} from "../belts"
import {
  Direction,
  axisSign,
  createRay,
  directionAxis,
  getRayPosition,
  isBeforeCmpOnRay,
  isBeforeOnRay,
  oppositeDirection,
  rayPosition,
  rayRelativeDirection,
  type Ray,
  type TilePosition,
} from "../geometry"
import { tryRegister } from "../metatable"
import { WorldOps, type World } from "../world"
import { Action, ActionError } from "./action"
import { LastBuiltEntity, step, type DragStepResult } from "./drag_state"
import { RaySense, senseMultiplier } from "./RaySense"
import { SmartBeltWorldView, type TileHistory } from "./world_view"

export interface ErrorHandler {
  handleError(position: TilePosition, error: ActionError): void
}

@tryRegister
export class LineDrag {
  forwardPlacement: number
  backwardPlacement: number
  furthestPlacementSense: RaySense = RaySense.Forward
  private constructor(
    public ray: Ray,
    private tier: BeltTier,
    private lastBuiltEntity: LastBuiltEntity | undefined,
    private overImpassable: RaySense | undefined,
    private lastPosition: number,
    private tileHistory: TileHistory | undefined,
    private lastEndTileHistory: TileHistory | undefined,
    startCoord: number,
  ) {
    this.forwardPlacement = startCoord
    this.backwardPlacement = startCoord
  }

  private static newDrag(
    world: World,
    errorHandler: ErrorHandler,
    tier: BeltTier,
    startPos: TilePosition,
    beltDirection: Direction,
    firstBeltDirection: Direction,
    allowFastReplace: boolean,
  ): LineDrag {
    const worldOps = new WorldOps(world)
    const canPlace = world.canPlaceOrFastReplace(
      startPos,
      beltDirection,
      allowFastReplace,
    )
    const tileHistory: TileHistory | undefined = canPlace
      ? [startPos, worldOps.beltConnectionsAt(startPos)]
      : undefined

    const ray = createRay(startPos, beltDirection)
    const startCoord = rayPosition(ray, startPos)

    let lastBuiltEntity: LastBuiltEntity | undefined
    if (canPlace) {
      const belt = worldOps.placeBelt(startPos, firstBeltDirection, tier)
      lastBuiltEntity = LastBuiltEntity.new(belt, startCoord)
    } else {
      errorHandler.handleError(startPos, ActionError.EntityInTheWay)
      lastBuiltEntity = undefined
    }

    return new LineDrag(
      ray,
      tier,
      lastBuiltEntity,
      undefined,
      startCoord,
      tileHistory,
      undefined,
      startCoord,
    )
  }

  static startDrag(
    world: World,
    errorHandler: ErrorHandler,
    tier: BeltTier,
    startPos: TilePosition,
    beltDirection: Direction,
  ): LineDrag {
    return LineDrag.newDrag(
      world,
      errorHandler,
      tier,
      startPos,
      beltDirection,
      beltDirection,
      true,
    )
  }

  interpolateTo(
    world: World,
    errorHandler: ErrorHandler,
    newPosition: TilePosition,
  ): void {
    const targetPos = rayPosition(this.ray, newPosition)
    const axisStep = axisSign(this.ray.direction)
    let iterations = 0

    while (isBeforeOnRay(this.ray, this.lastPosition, targetPos)) {
      if (iterations++ >= 100) {
        throw new Error("Too many iterations in interpolateTo (forward)")
      }
      const nextPos = this.lastPosition + axisStep
      this.doStep(world, errorHandler, nextPos)
    }

    while (isBeforeOnRay(this.ray, targetPos, this.lastPosition)) {
      if (iterations++ >= 100) {
        throw new Error("Too many iterations in interpolateTo (backward)")
      }
      const nextPos = this.lastPosition - axisStep
      this.doStep(world, errorHandler, nextPos)
    }
  }

  private doStep(
    world: World,
    errorHandler: ErrorHandler,
    nextPosition: number,
  ): void {
    const lastEntityPos =
      this.lastBuiltEntity !== undefined
        ? this.lastBuiltEntity.position
        : this.lastPosition
    const cmp = isBeforeCmpOnRay(this.ray, lastEntityPos, nextPosition)
    let raySense: RaySense
    if (cmp < 0) {
      raySense = RaySense.Forward
    } else if (cmp > 0) {
      raySense = RaySense.Backward
    } else {
      this.lastPosition = nextPosition
      return
    }
    const view = this.createWorldView(world, nextPosition, raySense)
    const result = step(this.lastBuiltEntity, this.overImpassable, view)
    this.applyStep(world, errorHandler, result, nextPosition, raySense)
    this.lastPosition = nextPosition
  }

  getRotationPivot(): [position: TilePosition, isBackward: boolean] {
    const furthestPos = this.furthestPlacementPos()
    return [
      getRayPosition(this.ray, furthestPos),
      this.furthestPlacementSense === RaySense.Backward,
    ]
  }

  rotate(
    world: World,
    errorHandler: ErrorHandler,
    cursorPos: TilePosition,
  ): [newDrag: LineDrag, ok: boolean] {
    const turnDirection = rayRelativeDirection(this.ray, cursorPos)
    if (turnDirection === undefined) {
      return [this, false]
    }

    const [pivot, backward] = this.getRotationPivot()
    const oldDirection = this.ray.direction
    const newBeltDirection = backward
      ? oppositeDirection(turnDirection)
      : turnDirection
    const firstBeltDirection = backward ? oldDirection : turnDirection

    const lastTileHistory =
      this.tileHistory !== undefined &&
      !backward &&
      this.furthestPlacementPos() === this.lastPosition
        ? this.tileHistory
        : undefined

    const newLineDrag = LineDrag.newDrag(
      world,
      errorHandler,
      this.tier,
      pivot,
      newBeltDirection,
      firstBeltDirection,
      false,
    )
    newLineDrag.lastEndTileHistory = lastTileHistory
    newLineDrag.interpolateTo(world, errorHandler, cursorPos)
    return [newLineDrag, true]
  }

  private applyStep(
    world: World,
    errorHandler: ErrorHandler,
    result: DragStepResult,
    nextPosition: number,
    raySense: RaySense,
  ): void {
    const [action, err] = result
    this.applyAction(world, errorHandler, action, nextPosition, raySense)

    if (err !== undefined) {
      const worldPos = getRayPosition(this.ray, nextPosition)
      errorHandler.handleError(worldPos, err)
    }
  }

  private storeTileHistory(position: number, world: World): void {
    const worldOps = new WorldOps(world)
    const worldPos = getRayPosition(this.ray, position)
    this.tileHistory = [worldPos, worldOps.beltConnectionsAt(worldPos)]
  }

  private updateFurthestPlacement(position: number, world: World): void {
    if (isBeforeOnRay(this.ray, this.forwardPlacement, position)) {
      this.forwardPlacement = position
      this.storeTileHistory(position, world)
      this.furthestPlacementSense = RaySense.Forward
    }
    if (isBeforeOnRay(this.ray, position, this.backwardPlacement)) {
      this.backwardPlacement = position
      this.storeTileHistory(position, world)
      this.furthestPlacementSense = RaySense.Backward
    }
  }

  private beforeEntityPlaced(position: number, world: World): void {
    this.updateFurthestPlacement(position, world)
  }

  private applyAction(
    world: World,
    errorHandler: ErrorHandler,
    action: Action,
    nextPosition: number,
    raySense: RaySense,
  ): void {
    const worldOps = new WorldOps(world)
    const worldPos = getRayPosition(this.ray, nextPosition)

    switch (action.type) {
      case "None":
        break

      case "PlaceBelt": {
        this.beforeEntityPlaced(nextPosition, world)
        const belt = worldOps.placeBelt(worldPos, this.ray.direction, this.tier)
        this.setLastBuiltEntity(LastBuiltEntity.new(belt, nextPosition))
        break
      }

      case "CreateUnderground": {
        const inputWorldPos = getRayPosition(this.ray, action.inputPos)
        const outputWorldPos = getRayPosition(this.ray, action.outputPos)

        worldOps.placeUndergroundBelt(
          inputWorldPos,
          this.ray.direction,
          raySense === RaySense.Forward,
          this.tier,
          false,
        )

        this.beforeEntityPlaced(action.outputPos, world)
        const outputEntity = worldOps.placeUndergroundBelt(
          outputWorldPos,
          this.ray.direction,
          raySense === RaySense.Backward,
          this.tier,
          true,
        )
        this.setLastBuiltEntity(
          LastBuiltEntity.new(outputEntity, action.outputPos),
        )
        break
      }

      case "ExtendUnderground": {
        const previousOutputWorldPos = getRayPosition(
          this.ray,
          action.lastOutputPos,
        )
        const newOutputWorldPos = getRayPosition(this.ray, action.newOutputPos)

        world.mine(previousOutputWorldPos)

        this.beforeEntityPlaced(action.newOutputPos, world)
        const entity = worldOps.placeUndergroundBelt(
          newOutputWorldPos,
          this.ray.direction,
          raySense === RaySense.Backward,
          this.tier,
          false,
        )
        this.setLastBuiltEntity(
          LastBuiltEntity.new(entity, action.newOutputPos),
        )
        break
      }

      case "IntegrateInputUnderground": {
        const ug = world.get(worldPos)
        if (!(ug instanceof UndergroundBelt)) {
          throw new Error("Expected UndergroundBelt at position")
        }

        if (ug.isInput !== (raySense === RaySense.Forward)) {
          world.flipUg(worldPos)
        }

        if (ug.tier != this.tier) {
          const view = this.createWorldView(world, nextPosition, raySense)
          if (canUpgradeUnderground(view, action.outputPos)) {
            world.upgradeUg(worldPos, this.tier)
          } else {
            LineDrag.reportError(
              errorHandler,
              ActionError.CannotUpgradeUnderground,
              worldPos,
            )
          }
        }

        const entity = world.get(worldPos)
        if (entity instanceof BeltConnectable) {
          this.setLastBuiltEntity(LastBuiltEntity.new(entity, nextPosition))
        }
        break
      }

      case "IntegrateOutputUnderground": {
        const entity = world.get(worldPos)
        if (entity instanceof BeltConnectable) {
          this.setLastBuiltEntity(LastBuiltEntity.new(entity, nextPosition))
        }
        break
      }

      case "IntegrateSplitter": {
        const entity = world.get(worldPos)
        if (!(entity instanceof Splitter)) {
          throw new Error("Expected Splitter at position")
        }

        if (
          this.tier.splitterName !== undefined &&
          entity.name !== this.tier.splitterName
        ) {
          world.upgradeSplitter(worldPos, this.tier.splitterName)
        }

        this.setLastBuiltEntity(LastBuiltEntity.new(entity, nextPosition))
        break
      }

      case "SetImpassable":
        this.overImpassable = action.raySense
        break

      case "ClearEntity":
        this.lastBuiltEntity = undefined
        this.overImpassable = undefined
        break
    }
  }

  private setLastBuiltEntity(entity: LastBuiltEntity): void {
    this.lastBuiltEntity = entity
    this.overImpassable = undefined
  }

  private static reportError(
    errorHandler: ErrorHandler,
    error: ActionError,
    worldPos: TilePosition,
  ): void {
    errorHandler.handleError(worldPos, error)
  }

  private furthestPlacementPos(): number {
    return this.furthestPlacementSense === RaySense.Forward
      ? this.forwardPlacement
      : this.backwardPlacement
  }

  private createWorldView(
    world: World,
    targetPos: number,
    relativeSense: RaySense,
  ): SmartBeltWorldView {
    const tileHistory: TileHistory[] = [
      ...(this.tileHistory ? [this.tileHistory] : []),
      ...(this.lastEndTileHistory ? [this.lastEndTileHistory] : []),
    ]
    const senseFurthestPos =
      relativeSense === RaySense.Forward
        ? this.forwardPlacement
        : this.backwardPlacement
    return new SmartBeltWorldView(
      world,
      tileHistory,
      this.ray,
      relativeSense,
      this.tier,
      targetPos,
      senseFurthestPos,
    )
  }
}

function canUpgradeUnderground(
  view: SmartBeltWorldView,
  outputPos: number,
): boolean {
  const inputPos = view.nextPosition()

  if (Math.abs(outputPos - inputPos) > view.tier.undergroundDistance) {
    return false
  }

  const start = Math.min(inputPos, outputPos) + 1
  const end = Math.max(inputPos, outputPos) - 1

  for (let pos = start; pos <= end; pos++) {
    const worldPos = getRayPosition(view.ray, pos)
    const entity = view.world.get(worldPos)

    if (entity instanceof UndergroundBelt) {
      const entityAxis = directionAxis(entity.direction)
      const rayAxis = directionAxis(view.ray.direction)

      if (entityAxis === rayAxis && entity.tier === view.tier) {
        return false
      }
    }
  }

  return true
}
